server/src/main/java/org/opentosca/toscana/plugins/cloudformation/visitor/PrepareModelNodeVisitor.java

```diff
@@ -0,0 +1,79 @@
+package org.opentosca.toscana.plugins.cloudformation.visitor;
+
+import java.security.SecureRandom;
+
+import org.opentosca.toscana.model.node.Compute;
+import org.opentosca.toscana.model.node.MysqlDatabase;
+import org.opentosca.toscana.model.node.RootNode;
+import org.opentosca.toscana.model.relation.RootRelationship;
+import org.opentosca.toscana.model.visitor.NodeVisitor;
+
+import com.scaleset.cfbuilder.core.Fn;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.jgrapht.Graph;
+import org.slf4j.Logger;
+
+import static org.opentosca.toscana.plugins.cloudformation.CloudFormationLifecycle.toAlphanumerical;
+import static org.opentosca.toscana.plugins.cloudformation.visitor.CloudFormationNodeVisitor.getCompute;
+
+public class PrepareModelNodeVisitor implements NodeVisitor {
+
+    protected static final String AWS_ENDPOINT_REFERENCE = "Endpoint.Address";
+    private static final int minPWLength = 8;
+    private static final String DEFAULT_USER = "root";
+    private static final Integer DEFAULT_PORT = 3306;
+    private final Logger logger;
+    private Graph<RootNode, RootRelationship> topology;
+
+    public PrepareModelNodeVisitor(Logger logger, Graph<RootNode, RootRelationship> topology) {
+        this.logger = logger;
+        this.topology = topology;
+    }
+
+    @Override
+    public void visit(MysqlDatabase node) {
+        logger.info("Prepare MysqlDatabase node {}.", node.getEntityName());
+        //if certain values aren't given, fill them
+        if (node.getPassword().isPresent()) {
+            //password needs to be at least 8 characters long
+            String password = node.getPassword().get();
+            if (password.length() < minPWLength) {
+                logger.warn("Database password to short, creating new random password");
+                node.setPassword(randomString(minPWLength));
+            }
+        } else {
+            logger.warn("No database password given, creating new random password");
+            node.setPassword(randomString(minPWLength));
+        }
+        if (!node.getUser().isPresent()) {
+            logger.warn("User not set, setting to default");
+            node.setUser(DEFAULT_USER);
+        }
+        if (!node.getPort().isPresent()) {
+            logger.warn("Database port not set, setting to default");
+            node.setPort(DEFAULT_PORT);
+        }
+
+        // check if Mysql is the only node hosted on his compute node
+        Compute compute = getCompute(node);
+        if (topology.incomingEdgesOf(compute).size() == 1) {
+            // means our dbms is the only one hosted on this compute
```

Comment:
This is incorrect. The graph's edges are a Set of all possible Relationships. HostedOn is only one of them.
A quick fix could be:
```java
if ( topology.incomingEdgesOf(compute).stream().filter(p -> p instanceof HostedOn).collect(Collectors.toSet()).size() == 1)
```
Didn't try that code, though.

Reply:
Actually I'm pretty sure this is correct. Sure it is a Set of all possible relationships but I already know that the mysqlDatabase node I am currently visiting is hosted on this specific compute (or mysqlDatabase is hosted on mysqlDbms is hosted on this compute node).
That means if the compute node has exactly one relationship it is that hosted on relation.

Reply:
Imagine the tosca modeller has written a template that defines a timing dependency (DependsOn) from node 'otherNode' to 'server'.
Now, 
```java
if (topology.incomingEdgesOf(compute).size() == 1) {
```
..will evaluate to false, altough mysql dbms node is still the only node hosted on the server node.

Reply:
True